src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -428,29 +443,324 @@ Shard::valFetch(uint256 const& hash)
+            app_.config().getSize(siNodeCacheAge)}};
+
+        pCache_ = std::make_shared<PCache>(name, sz, age, stopwatch(), j_);
+        nCache_ = std::make_shared<NCache>(name, stopwatch(), sz, age);
+    }
+    else
+    {
+        auto const sz {Config::getSize(siNodeCacheSize, 0)};
+        pCache_->setTargetSize(sz);
+        nCache_->setTargetSize(sz);
+
+        auto const age {std::chrono::seconds{
+            Config::getSize(siNodeCacheAge, 0)}};
+        pCache_->setTargetAge(age);
+        nCache_->setTargetAge(age);
+    }
+}
+
+bool
+Shard::initSQLite()
+{
+    Config const& config {app_.config()};
+    DatabaseCon::Setup setup;
+    setup.startUp = config.START_UP;
+    setup.standAlone = config.standalone();
+    setup.dataDir = dir_;
+
+    try
+    {
+        if (complete_)
         {
-            JLOG(j_.error()) <<
-                "unknown error. hash " << hash;
+            using namespace boost::filesystem;
+
+            // Remove WAL files if they exist
+            for (auto const& p : recursive_directory_iterator(dir_))
+            {
+                if (is_regular_file(p) &&
+                    boost::iends_with(extension(p), "-wal"))
+                {
+                    // Closing the session forces a checkpoint
+                    if (!lgrSQLiteDB_)
+                    {
+                        lgrSQLiteDB_ = std::make_unique <DatabaseCon>(
+                            setup,
+                            LgrDBName,
+                            LgrDBPragma,
+                            LgrDBInit);
+                    }
+                    lgrSQLiteDB_->getSession().close();
+
+                    if (!txSQLiteDB_)
+                    {
+                        txSQLiteDB_ = std::make_unique <DatabaseCon>(
+                            setup,
+                            TxDBName,
+                            TxDBPragma,
+                            TxDBInit);
+                    }
+                    txSQLiteDB_->getSession().close();
+                    break;
+                }
+            }
+
+            lgrSQLiteDB_ = std::make_unique <DatabaseCon>(
+                setup,
+                LgrDBName,
+                CompleteShardDBPragma,
+                LgrDBInit);
+            lgrSQLiteDB_->getSession() <<
+                boost::str(boost::format("PRAGMA cache_size=-%d;") %
+                kilobytes(Config::getSize(siLgrDBCache, 0)));
+
+            txSQLiteDB_ = std::make_unique <DatabaseCon>(
+                setup,
+                TxDBName,
+                CompleteShardDBPragma,
+                TxDBInit);
+            txSQLiteDB_->getSession() <<
+                boost::str(boost::format("PRAGMA cache_size=-%d;") %
+                kilobytes(Config::getSize(siTxnDBCache, 0)));
         }
+        else
+        {
+            // The incomplete shard uses a Write Ahead Log for performance
+            lgrSQLiteDB_ = std::make_unique <DatabaseCon>(
```

Comment:
nit: these two blocks are pretty duplicated aside from a few pieces like `setupCheckpointing` and the Pragma parameter passed in during construction.